pkg/vault/contracts/AddUnbalancedLiquidityViaSwapRouter.sol

```diff
@@ -0,0 +1,228 @@
+
+    function queryAddUnbalancedLiquidityViaSwapHook(
+        AddLiquidityAndSwapHookParams calldata params
+    ) external nonReentrant onlyVault returns (uint256[] memory amountsIn, uint256 swapAmountCalculated) {
+        (amountsIn, , ) = _queryAddLiquidityHook(params.addLiquidityParams);
+        swapAmountCalculated = _querySwapHook(params.swapParams);
+    }
+
+    /***************************************************************************
+                                   Others functions
+    ***************************************************************************/
+
+    // Required to avoid stack-too-deep in the caller.
+    function _buildAddLiquidityParams(
+        address pool,
+        address sender,
+        uint256 deadline,
+        bool wethIsEth,
+        AddLiquidityProportionalParams calldata addLiquidityParams,
+        SwapParams memory swapParams
+    ) private pure returns (AddLiquidityAndSwapHookParams memory params) {
+        return
+            AddLiquidityAndSwapHookParams({
+                addLiquidityParams: AddLiquidityHookParams({
+                    sender: sender,
+                    pool: pool,
+                    maxAmountsIn: addLiquidityParams.maxAmountsIn,
+                    minBptAmountOut: addLiquidityParams.exactBptAmountOut,
+                    kind: AddLiquidityKind.PROPORTIONAL,
+                    wethIsEth: wethIsEth,
+                    userData: addLiquidityParams.userData
+                }),
+                swapParams: SwapSingleTokenHookParams({
+                    sender: sender,
+                    kind: swapParams.kind,
+                    pool: pool,
+                    tokenIn: swapParams.tokenIn,
+                    tokenOut: swapParams.tokenOut,
+                    amountGiven: swapParams.amountGiven,
+                    limit: swapParams.limit,
+                    deadline: deadline,
+                    wethIsEth: wethIsEth,
+                    userData: swapParams.userData
+                })
+            });
+    }
+
+    function _buildSwapExactInParams(SwapExactInParams calldata swapParams) private pure returns (SwapParams memory) {
+        return
+            SwapParams({
+                tokenIn: swapParams.tokenIn,
+                tokenOut: swapParams.tokenOut,
+                kind: SwapKind.EXACT_IN,
+                amountGiven: swapParams.exactAmountIn,
+                limit: swapParams.minAmountOut,
+                userData: swapParams.userData
+            });
+    }
+
+    function _buildSwapExactOutParams(SwapExactOutParams calldata swapParams) private pure returns (SwapParams memory) {
+        return
+            SwapParams({
+                tokenIn: swapParams.tokenIn,
+                tokenOut: swapParams.tokenOut,
+                kind: SwapKind.EXACT_OUT,
+                amountGiven: swapParams.exactAmountOut,
+                limit: swapParams.maxAmountIn,
+                userData: swapParams.userData
+            });
+    }
```

Comment:
This only adds the kind right?
Perhaps you can build `_buildAddLiquidityParams` directly from `SwapExactInParams` or `SwapExactOutParams` + kind? Or does that cause stack too deep?

(just to clarify, this looks fine as is and there's no need to make changes; I'm just curious).